### PR TITLE
Added Neural Network Architecture Automation

### PR DESCRIPTION
**Pull request by samysspace**
After a considerable amount of investigation, I realized that TPot does not have any good way of adding in custom regression or classifier classes even when wrapped by a SciKit Learn BaseEstimator with RegressionMixin or ClassifierMixin (see https://github.com/EpistasisLab/tpot/issues/323, https://github.com/EpistasisLab/tpot/issues/407, and https://github.com/EpistasisLab/tpot/issues/602, all of which present very hacky solutions that do not mesh well with the current configuration dictionaries). My solution was that I built a best_model function which simulates what TPot does by ensembling various optimizers, initializers, and activation functions to produce the best NN architecture for the given training data (measured by cross-validation score). This has been integrated with the fit method of the neural network adaptor and works pretty well.

### CODE REVIEW

**Comment by ardunn** [comment]
Hey Samy, looking better! There are a few issues though:

1. there seem to be some conflicting files^^^?
2. Make sure your adaptor inherits all of the required properties/methods from its base class, namely `AutoMLAdaptor`. I noticed there is no `_best_pipeline`/`best_pipeline` attr?
3. Could you try out this adaptor on a full MatPipe run? Dataset doesn't have to be huge and results don't have to be good, just want to make sure we won't run into weird problems later on. For example try running the following matpipe (may need to be adjusted):

```python
from matminer.datasets_data_retrieval import load_dataset

learner = NeuralNetworkAdatptor()
autofeaturizer = AutoFeaturizer(composition_col="formula")
pipe = MatPipe(learner=learner, autofeaturizer=autofeaturizer)
df = load_dataset("elastic_tensor_2015")
predictions = pipe.benchmark(df[:-100], "K_VRH")
pipe.digest()
pipe.save("somefile.p")
pipe.load("somefile.p")
pipe.predict(df[-100:], "K_VRH")
```

This is basically running through all the motions we'll need NNAdaptor to handle. Let me know what problems you run into (if any)

**Comment by ardunn** [comment]
@samysspace btw don't worry if next tests don't pass, they are currently failing bc of pip-related matminer bug :/

**Comment by ardunn** [comment]
@samysspace actually as of now new pushes you make should have all tests pass :)

**Comment by samysspace** [comment]
Hey @ardunn, sorry about that - Everything looks good to go at the moment. I'm pushing what I have right now which appears to be giving me functional results without errors/passing all my tests but I'll be testing continuously over the weekend to make sure since I recently refactored my code.

**Comment by ardunn** [comment]
@samysspace sure thing, thanks for the update! Build is probably failing for reasons unrelated to this PR, trying to fix at the moment.

**Comment by ardunn** [comment]
@samysspace nvm, tests are good now. when they pass, I'll do a code review and we'll merge if all good

**Comment by samysspace** [comment]
After doing some extensive testing on both the AutoML models and the neural network adaptor, it appears that the neural network adaptor lags behind the AutoML adaptor in terms of prediction accuracy. I conducted most of my tests on the elastic_tensor_2015 dataset, trying to predict the K_VRH of the compounds. While the neural network averaged a root squared error of ~202, the AutoML adaptor averaged an RMSE of ~168. In similar tests done using other metrics such as MAE (to account for outliers) and R^2 score, the AutoML adaptor beats out the neural network as well. A few bright sides to this situation though: 
1) The neural network adaptor outperforms a randomly selected neural network architecture 95% of the time showing that the genetic algorithm optimizer does indeed optimize the architecture. 
2) For the purposes of testing, I severely limited the number of generations the algorithm is allowed to go through, as well as limiting other factors to speed up testing a little bit. This means that the neural network adaptor completes in about 5 minutes whereas the TPot adaptor takes about 30-45 minutes to run locally. This shows that the neural network does provide a somewhat accurate model in a very short amount of time.
3) I found a small bug in my genetic algorithm (specifically, which models are being carried over to the next generation) over the course of testing. By hopefully changing this algorithm slightly and making a few other slight modifications, I can boost accuracy. This should be done by next Monday.
4) While TPot performs better in terms of RMSE by a wide margin, the MAE values are relatively closer seemingly indicating the presence of dominating outliers. I will continue to look into this and just like Point #3, I will get this done by next Monday.

**Comment by samysspace** [comment]
Sounds good. See you tomorrow.

**Comment by ardunn** [comment]
1. Do the same test on "mp_all" dataset with [["composition", "gap pbe"]] columns, "gap pbe" is target.
2. Do a test on elastic tensor 2015 (all 1,811 samples) with [["composition", "K_VRH"]] with the NN. Then record the time and set the TPOTAdaptor time to that value (i.e., so they run for the same amount of time, e.g., 30 min). Then compare them.

If you can, each test should be done with **Nested CV**. If this is too time consuming, just make sure you have a good randomly split test set **which is the same between methods** (so TPOT and NN have the same test!). Easiest way to do this is with a sklearn KFold object with the random_state argument specified, e.g., `KFold(n_splits=5, random_state=02062019, shuffle=True)`

**Comment by samysspace** [comment]
Updated the genetic algorithm to have a more streamlined, or readable, interface (currently in gen_solver_v2.py) and fixed a slight bug I found in v1. I tested it with a consistent 5-fold object using a fixed random seed on the full elastic tensor 2015 dataset (since nested CV was a bit too intensive for my machine), predicting K_VRH. The genetic algorithm used 10 generations with a count of 15 nets per generation and each network's fitness was calculated using cross-validation scores using a 4-fold sklearn object. It consistently hovered around 16 for MAE whereas TPot, at least on my machine using the same amount of time as the genetic algorithm, seems to hover around 12. This shows promise for much larger datasets such as "mp all". Unfortunately, running mp all on my machine seems to stall out and freeze or crash my IDE. I still have not been granted access to Lawrencium yet either. Furthermore, it appears the recently pushed changes to the adaptors.py file has rendered my testing interface no longer runnable (I just refetched from upstream today and it seems like a bunch of critical components have been replaced) so I will try to understand the changes tomorrow at our weekly meeting so I can get the tests back up and running.

As for future steps, all I have left is testing/waiting for a Lawrencium account so feel free to assign me to any other pertinent issues within the project.

**Comment by ardunn** [comment]
@samysspace 

Ok sounds good, thanks for update. Excited to see mp-all results. We'll talk more tomorrow but here is a summary of what we should do before I'll merge this PR. 

0. Get the mp-all results. Hopefully NERSC gets back to you soon??

1. Adaptors inheritable has just changed name to DFMLTransformer. NNAdaptor can inherit from that now. DFMLTransformer is an abstract base class, **which means you need to implement all its methods and abstract properties to not get errors upon instantiation of NNAdaptor** (for example of this, just see the new TPOTAdaptor, or for a simpler version, the SingleModelAdaptor). For example, you need to define a `fitted_pipeline` property (see SingleModelAdaptor/TPOTAdaptor for example). 

2. Make sure to add the @check_fitted and @set_fitted decorators to any NNAdaptor methods that require a fit (e.g., `.transform`) or set a fit (e.g., `.fit`). Also add @log_progress to any DFMLAdaptor methods, like fit and transform. Check out the TPOTAdaptor for an example of this (shouldn't throw any errors you weren't getting before)

3. The full tests now run automatically everytime there is a release of automatminer. Change your tests to be as lean/fast as you can make them (the nets can be very shallow, number of training points very low, etc.) if they are taking more than 5 min to run locally. Then change 
```python
@unittest.skipIf("CI" in os.environ.keys(), "Test too intensive for CircleCI.")
```
To
```python
@unittest.skipIf(int(os.environ.get("SKIP_INTENSIVE", 0)),
                 "Test too intensive for CircleCI commit builds.")
```
When these are done, I'll run a full test.

BTW I can go ahead and do items 1-3 tomorrow if you give me push access to ur branch. Then tests should pass

**Comment by ardunn** [comment]
@samysspace I'll be closing this PR and resubmitting with modifications, but I am still going to leave a code review, which will help you improve your python skills for PRs in the future

File: automatminer/automl/gen_evolve_v2.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        self.mutate_chance = mutate_chance
+        self.retain = retain
+        self.X = X
+        self.y = y
+        self.Model = model
+        ACTIVATION = ["sigmoid", "tanh", "relu", "elu"]
+        OPTIMIZERS = ["sgd", "rmsprop", "adagrad", "adadelta", "nadam", "adamax", "adam"]
+        self.NN_PARAM_CHOICES = {"units": range(1, 1000), "hidden_layer_sizes": range(1, 3), "optimizer": OPTIMIZERS,
+                            "activation": ACTIVATION}
+        pop = self.create_population(count)
```
Comment (ardunn) [inline automatminer/automl/gen_evolve_v2.py]: As far as I can tell, you only run create_population once in this entire class, and it is not used as a method anywhere else in the code. It doesn't need to be its own method if it is only being used in init, and it is pretty cheap to run as far as i can tell.

On the other hand, if you expect create_population to be repeatedly used (i.e., to generate new populations using the same object) or it is expensive enough to be its own operation, it's ok to have its own method. But that does not seem like the use case here.

File: automatminer/automl/gen_evolve_v2.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        for _ in range(10):
+            pop = self.evolve(pop, X, y)
```
Comment (ardunn) [inline automatminer/automl/gen_evolve_v2.py]: Just initializing the class should probably not run any super expensive operations. We are just initializing the class, after all - this seems to be running the entire GA (including training etc.!!) right here in __init__.

This is the opposite problem we have with self.create_population. This probably should have its own method, e.g., `optimize`.

File: automatminer/automl/gen_evolve_v2.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+        self.model_win = self.best_model(pop)
+
+    def create_population(self, count):
```
Comment (ardunn) [inline automatminer/automl/gen_evolve_v2.py]: if possible, I'd try to use more descriptive variable names than `count`. For example, `pop_size`, which is more reflective of what the variable actually is: population size.

File: automatminer/automl/keras_wrapper.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
+from __future__ import unicode_literals
```
Comment (ardunn) [inline automatminer/automl/keras_wrapper.py]: We are working in Python3.6+ exclusively. If I remember correctly, unicode_literals are standard in python3 and this import is not needed.

File: automatminer/automl/keras_wrapper.py [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+from gen_evolve_v2 import NeuralNetOptimizer
+import numpy as np
+import random
+
+_classifier_modes = {'classifier', 'classification', 'classify'}
+
+_regressor_modes = {'regressor', 'regression', 'regress'}
+
+class NnWrapper(sklearn.base.BaseEstimator, sklearn.base.RegressorMixin, sklearn.base.ClassifierMixin):
+    """Wrapper for Keras feed-forward neural network for classification to enable scikit-learn grid search"""
```
Comment (ardunn) [inline automatminer/automl/keras_wrapper.py]: gonna need some more docs than this

File: automatminer/automl/keras_wrapper.py [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+import keras.callbacks
+from sklearn.datasets import load_boston
+from sklearn.model_selection import KFold, cross_val_score
+from keras.wrappers.scikit_learn import KerasRegressor, KerasClassifier
+import pandas as pd
+from gen_evolve_v2 import NeuralNetOptimizer
+import numpy as np
+import random
+
+_classifier_modes = {'classifier', 'classification', 'classify'}
```
Comment (ardunn) [inline automatminer/automl/keras_wrapper.py]: Don't copy and paste constants like this.

Either: 
1. Import them from the other module.
2. Move them to a common file and import them to both modules.

File: automatminer/automl/keras_wrapper.py [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+        self.units = units
+
+        if self.use_maxout:
+            self.use_maxnorm = True
+
+        self.model_ = None
+        self.estimator = None
+        self.classifier = None
+        self.kfold_splits = kfold_splits
+        if mode in _classifier_modes:
```
Comment (ardunn) [inline automatminer/automl/keras_wrapper.py]: So if I set `mode=dnasjdasd8as8das8d8sa8`, or any other string or object which is not in _classifier_modes, this adaptor will set the mode to "regression". This can be confusing.

File: automatminer/automl/keras_wrapper.py [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+            self.mode = "classification"
+        else:
+            self.mode = "regression"
+        self.fitted_pipeline_ = None
+
+    def getModel(self):
```
Comment (ardunn) [inline automatminer/automl/keras_wrapper.py]: don't use camelcase for methods. See Google python code style standards

Also, why is this method needed? You should be able to just call self.model_ whenever you need to?

File: automatminer/automl/keras_wrapper.py [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+        return self.model_
```
Comment (ardunn) [inline automatminer/automl/keras_wrapper.py]: Trailing underscores in python are used to avoid conflicts with keywords (see https://dbader.org/blog/meaning-of-underscores-in-python). If you want this to be private, use self._model.

File: automatminer/automl/keras_wrapper.py [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+
+    def fit(self, X, y, **kwargs):
```
Comment (ardunn) [inline automatminer/automl/keras_wrapper.py]: Needs docs. Specify what the kwargs can be and where to find more info.

File: automatminer/automl/keras_wrapper.py [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+            model.compile(loss="binary_crossentropy", optimizer=self.optimizer, metrics=["accuracy"])
+        else:
+            model.compile(loss="mse", optimizer=self.optimizer, metrics=["mse"])
+
+        # batches as per configuration
+        for num_iterations, batch_size in self.batch_spec:
+            callbacks = None
+            validation_split = 0.0
+            if self.stop_early and batch_size > 0:
+                callbacks = [EarlyStopping(monitor='val_loss', patience=20, verbose=1)]
```
Comment (ardunn) [inline automatminer/automl/keras_wrapper.py]: Ur missing an import statement for this early stoppage. It will throw an importerror if it ever gets to this point.

File: requirements.txt [hunk 12]
```diff
@@ -2,4 +2,6 @@ matminer==0.5.0
 xlrd==1.1.0
 xgboost==0.72.1
 tpot==0.9.5
-skrebate==0.6
\ No newline at end of file
+skrebate==0.6
+Keras==2.2.4
+tensorflow==0.12.1
```
Comment (ardunn) [inline requirements.txt]: Non existent version of tensorflow. I think the one you want is 1.12.0

**Comment by ardunn** [comment]
Also as a general comment, there are not enough tests. **Each** class should have an individual test. Otherwise when something breaks it becomes very hard to figure out where and how it broke.
